lando/common/commands.py

```diff
@@ -0,0 +1,300 @@
+import os
+import json
+import subprocess
+import datetime
+import logging
+import codecs
+import tempfile
+from lando.exceptions import JobStepFailed
+from ddsc.config import LOCAL_CONFIG_ENV as DDSCLIENT_CONFIG_ENV, Config as DukeDSConfig
+
+RUN_CWL_COMMAND = "cwltool"
+RUN_CWL_OUTDIR_ARG = "--outdir"
+JOB_STDERR_OUTPUT_MAX_LINES = 100
+
+
+class StageDataTypes(object):
+    URL = "url"
+    WRITE = "write"
+    DUKEDS = "DukeDS"
+
+
+def read_file(file_path):
+    """
+    Read the contents of a file using utf-8 encoding, or return an empty string
+    if it does not exist
+    :param file_path: str: path to the file to read
+    :return: str: contents of file
+    """
+    try:
+        with codecs.open(file_path, 'r', encoding='utf-8', errors='xmlcharrefreplace') as infile:
+            return infile.read()
+    except OSError as e:
+        logging.exception('Error opening {}'.format(file_path))
+        return ''
+
+
+class StepProcess(object):
+    def __init__(self, command, stdout_path, stderr_path, env=None):
+        self.command = command
+        self.env = env
+        self.stdout_path = stdout_path
+        self.stderr_path = stderr_path
+        # properties filled in by run method
+        self.return_code = None
+        self.started = None
+        self.finished = None
+
+    def run(self):
+        self.started = datetime.datetime.now()
+        # Configure the subprocess to write stdout and stderr directly to files
+        logging.info('Running command: {}'.format(' '.join(self.command)))
+        logging.info('Redirecting stdout > {},  stderr > {}'.format(self.stdout_path, self.stderr_path))
+        stdout_file = open(self.stdout_path, 'w')
+        stderr_file = open(self.stderr_path, 'w')
+        try:
+            self.return_code = subprocess.call(self.command, env=self.env, stdout=stdout_file, stderr=stderr_file)
+        except OSError as e:
+            logging.error('Error running subprocess %s', e)
+            error_message = "Command failed: {}".format(' '.join(self.command))
+            raise JobStepFailed(error_message, e)
+        finally:
+            if stdout_file:
+                stdout_file.close()
+            if stderr_file:
+                stderr_file.close()
+            self.finished = datetime.datetime.now()
+
+    def total_runtime_str(self):
+        """
+        Returns a string describing how long the command took.
+        :return: str: "<number> minutes"
+        """
+        elapsed_seconds = (self.finished - self.started).total_seconds()
+        return "{} minutes".format(elapsed_seconds / 60)
+
+
+class BaseCommand(object):
+    @staticmethod
+    def write_json_file(filename, data):
+        with open(filename, 'w') as outfile:
+            outfile.write(json.dumps(data))
+
+    def run_command(self, command, env=None, stdout_path=None, stderr_path=None):
+        stdout_path, cleanup_stdout_path = self._create_temp_filename_if_none(stdout_path)
```

Comment:
The changes here re-used the logic that writes stdout and stderr to files when running a process. For running a workflow(eg. cwltool) we save stdout and stderr into particular locations that are included in the output project. For operations like staging data and uploading the results we do not currently keep the output. 

We could have all commands specify filenames for stdout/stderr. Or just copy the NamedTemporaryFile to the output destination when the command is run workflow.

Reply:
My point is that there's places where the code is deliberate for obvious reasons (e.g. running a workflow and redirecting the output to a named file) , and places where it's deliberate but not obvious. It makes sure the output is *always* redirected and does its own cleanup on temporary files there. My feedback is to provide the reasons for these decisions. 

As a prompt, you might answer the questions about why, in the case where the stdout/stderr paths are provided as None, why

1. the streams need to be redirected at all
2. the files are manually deleted rather than letting NamedTemporaryFile do its own cleanup

I recognize the answer lies in `_handle_failed_process()` but that's a couple layers removed from `run_command()`

Reply:
I reworked the logic to make it clearer how these files are being used and added some comments.